### PR TITLE
Corda-3710 Enable the use of AMQP for RPC requests/responses

### PR DESCRIPTION
**Pull request by djc**
While the AMQP protocol is enabled for the RPC queues in Corda's Artemis setup, in my testing I found that RPC calls over AMQP do not actually work. This PR makes the minimal changes required to make it work, tested with Artemis 2.6.2 (currently in use) and 2.11.0 (the newest version).

The permission changes from the first commit were briefly discussed with @mikehearn on Slack a few months ago, who said that the use of durable queues for at least some kinds of RPC calls would make sense.

This notably does not include testing of the AMQP path in the Corda test suite. I currently have manual tests as part of work on [corda-rpc](https://github.com/djc/corda-rpc/) which I will keep running. I'm not sure it's economical for me to write Kotlin-based AMQP tests, but I don't mind being on the hook for "maintaining" the AMQP support in the RPC subsystem for the foreseeable future. (That stuff didn't change much on the Artemis side in the two-year window between 2.6.2 and 2.11.0 gives me good hope that this shouldn't be too onerous.)

I hereby certify that my contribution is in accordance with the Developer Certificate of Origin (https://developercertificate.org/).

### CODE REVIEW

**Comment by davidrapacchiale** [comment]
Thanks @djc for your contribution. I have raised https://r3-cev.atlassian.net/browse/CORDA-3710 so that the Engineering team can progress the discussion directly on the jira ticket.

**Comment by djc** [comment]
I don't think I have access to that issue. Can someone grant me access?

**Review by mnesbit — CHANGES_REQUESTED** [review]
I am fine with this in principle, but we would absolutely need a simple AMQP client being created and some integration tests to confirm stable operations. Also all existing RPC via the standard client must not break, even for older 4.x rpc clients!

Please note, we have always regarded the RPC as a locally versioned protocol and thus we may change the way things work to enhance performance and reliability. For instance I am keen to remove the automatic queue/session detection and move to a more typical publish to well known topic to establish a session topic, then use heartbeats to manage the session. This would greatly facilitate out-of process and HA RPC support and let us produce re-connectable observable streams. The durable queue approach doesn't  lead into that and there is no clear way to deduplicate requests, or re-associate across to a failover node with this approach.

**Comment by davidrapacchiale** [comment]
Our Corda Jira project is OS, @djc. You should be able to access the ticket after signing up in Jira

File: node/src/main/kotlin/net/corda/node/services/rpc/RpcBrokerConfiguration.kt [hunk 1]
```diff
@@ -55,6 +55,8 @@ internal class RpcBrokerConfiguration(baseDirectory: Path, maxMessageSize: Int,
             "${RPCApi.RPC_CLIENT_QUEUE_NAME_PREFIX}.$username.#" to setOf(nodeInternalRole, restrictedRole(
                     "${RPCApi.RPC_CLIENT_QUEUE_NAME_PREFIX}.$username",
                     consume = true,
+                    createDurableQueue = true,
+                    deleteDurableQueue = true,
```
Comment (mnesbit) [inline node/src/main/kotlin/net/corda/node/services/rpc/RpcBrokerConfiguration.kt:63]: Please add a comment to say that the ability to create durable queue's is only experimental

**Comment by djc** [comment]
> Our Corda Jira project is OS, @djc. You should be able to access the ticket after signing up in Jira

While logged in, when I click your link to https://r3-cev.atlassian.net/browse/CORDA-3710 I get redirected to https://r3-cev.atlassian.net/servicedesk/customer/portals. (I'm not sure what the open source status has to do with access control for the Jira project.)

File: node/src/main/kotlin/net/corda/node/services/rpc/RpcBrokerConfiguration.kt (hunk 1 above)
Comment (djc) [reply node/src/main/kotlin/net/corda/node/services/rpc/RpcBrokerConfiguration.kt:63]: Done.

**Comment by djc** [comment]
> The durable queue approach doesn't lead into that and there is no clear way to deduplicate requests, or re-associate across to a failover node with this approach.

Note that I'm not interested in durable queue support per se. It's just that, with the current implementation of the `RPCServer` combined with the internals of how Artemis implements link acceptance, there does not seem to be a way to make RPC over AMQP work without durable queues (if you haven't yet, please review the commit messages for the individual commits for more details).

**Comment by mnesbit** [comment]
Yes I do understand that and indeed the lack of queue control (except simple dynamic) in AMQP1.0 is why we are still using CORE in various ways. I think your change is perfectly fine as it stands.

Would be good to build out the integration test though

**Comment by djc** [comment]
> I am fine with this in principle, but we would absolutely need a simple AMQP client being created and some integration tests to confirm stable operations.

Have you got any pointers to the right APIs to use to implement an AMQP client? I found `net/corda/nodeapi/internal/protonwrapper/netty/AMQPClient.kt` but it seems to be fairly low-level and more geared towards the needs of inter-node RPC (including the use of TLS rather than PLAIN SASL authentication).

> Also all existing RPC via the standard client must not break, even for older 4.x rpc clients!

I'm not sure how I would test this -- but I'ved tried to make sure that these changes are carefully scoped to have no effect on normal (Core) RPC flows.

**Comment by mnesbit** [comment]
Yes the AMQPClient has evolved to be tied to peer to peer communications more than I had originally hoped. I was hoping you already had some sort of client that could be used as a testCompile only dependency here. All the projects I have seen use proton-j under the hood e.g. https://qpid.apache.org/components/jms/index.html so see what is the simplest client possible? ArtemisMQ examples also include a tiny demo https://github.com/apache/activemq-artemis/blob/master/examples/protocols/amqp/queue/src/main/java/org/apache/activemq/artemis/jms/example/AMQPQueueExample.java

**Comment by djc** [comment]
As explained in the initial description, I use https://github.com/djc/corda-rpc/. I'm not sure this is the kind of thing you'd want in your Corda integration tests. I'll take a look at the Artemis example.

**Comment by vkolomeyko** [comment]
Hello Dirkjan (@djc ),
I was asked to help finishing your work such that it will be ready to ship with Corda 4.6.
I hope you do not mind me pushing into your fork/branch.
If this is a problem - please let me know and I will create my own branch in main Corda repo.

**Comment by djc** [comment]
@vkolomeyko thanks for the help, pushing into my branch is fine!

**Comment by vkolomeyko** [comment]
@mnesbit - I did more experiments this morning trying to perform an RPC call via AMQP JMS.
It does not work due to validation JMS layer performs for property names, see exception below.

Of course, it is possible to change string constants, but this will break compatibility with existing clients.

```
Identifier contains invalid JMS identifier character '-': 'rpc-id' 
java.lang.IllegalArgumentException: Identifier contains invalid JMS identifier character '-': 'rpc-id' 
	at org.apache.qpid.jms.message.JmsMessagePropertySupport.checkIdentifierLetterAndDigitRequirements(JmsMessagePropertySupport.java:101)
	at org.apache.qpid.jms.message.JmsMessagePropertySupport.checkPropertyNameIsValid(JmsMessagePropertySupport.java:63)
	at org.apache.qpid.jms.message.JmsMessagePropertyIntercepter.setProperty(JmsMessagePropertyIntercepter.java:724)
	at org.apache.qpid.jms.message.JmsMessage.setObjectProperty(JmsMessage.java:348)
	at org.apache.qpid.jms.message.JmsMessage.setStringProperty(JmsMessage.java:433)
	at net.corda.node.amqp.jms.AmqpRpcTests$Companion.mapTo(AmqpRpcTests.kt:52)
	at net.corda.node.amqp.jms.AmqpRpcTests$Companion.mapTo(AmqpRpcTests.kt:47)
	at net.corda.node.amqp.jms.AmqpRpcTests$Companion.access$mapTo(AmqpRpcTests.kt:35)
	at net.corda.node.amqp.jms.AmqpRpcTests$RPC over AMQP$1.invoke(AmqpRpcTests.kt:84)
	at net.corda.node.amqp.jms.AmqpRpcTests$RPC over AMQP$1.invoke(AmqpRpcTests.kt:33)
	at net.corda.testing.node.internal.DriverDSLImplKt.genericDriver(DriverDSLImpl.kt:1199)
	at net.corda.testing.driver.Driver.driver(Driver.kt:185)
	at net.corda.node.amqp.jms.AmqpRpcTests.RPC over AMQP(AmqpRpcTests.kt:59)
```

File: node/src/integration-test/kotlin/net/corda/node/amqp/jms/AmqpRpcTests.kt [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+                val message = session.createBytesMessage()
+                message.apply {
+                    jmsReplyTo = replyQueue
+                    setIntProperty(TAG_FIELD_NAME, RPC_REQUEST)
+                    replyId.mapTo(this)
+                    sessionId.mapTo(this)
+                    setStringProperty(METHOD_NAME_FIELD_NAME_ALT, "currentNodeTime")
+                    val serialisedArguments = emptyList<Unit>().serialize(context = AMQP_RPC_CLIENT_CONTEXT)
+                    writeBytes(serialisedArguments.bytes)
+                    setStringProperty(HDR_VALIDATED_USER, rpcUser.username) // TODO: We can pretend to be someone else and elevate our permissions
```
Comment (vkolomeyko) [inline node/src/integration-test/kotlin/net/corda/node/amqp/jms/AmqpRpcTests.kt:91]: @jamesbr3 and @mnesbit - we might need to discuss this to assess possible security implication.

File: node/src/main/kotlin/net/corda/node/services/rpc/RpcBrokerConfiguration.kt (hunk 1 above)
Comment (vkolomeyko) [reply node/src/main/kotlin/net/corda/node/services/rpc/RpcBrokerConfiguration.kt:63]: This change is necessary as `org.apache.activemq.artemis.protocol.amqp.broker.AMQPSessionCallback#queueQuery` creates queue with hardcoded value for `durable` as `true`.

**Review by vkolomeyko — DISMISSED** [review]

**Review by vkolomeyko — APPROVED** [review]

**Comment by rick-r3** [comment]
Unfortunately the outcome of this is that enabling the AMQP route allows a bypass of the current RPC security mechanisms.  We'd have to overhaul security of RPC first I think, before we can enable this.

So we can prioritise it, it would be good to hear what your use-case for this is?

**Comment by djc** [comment]
In general, I want to be able to access Corda RPC from non-JVM languages. In particular, I'm working on a web application backend application in Rust that I expect will want to make Corda RPC calls in the not too distant future. Given that you've already postponed this out to 4.6, deprioritizing this now would be a bit sad. If the problem here is that Artemis doesn't properly set/validate `HDR_VALIDATED_USER` for AMQP messages, perhaps the right solution is to implement that in Artemis (which seems straightforward), then adopt their next release into Corda 4.6?

**Comment by vkolomeyko** [comment]
Re: perhaps the right solution is to implement that in Artemis
This has already been raised with Artemis, please see: https://issues.apache.org/jira/browse/ARTEMIS-2751
They are unlikely to be making such an improvement as they claim that AMQP message is meant to be immutable.

**Comment by djc** [comment]
@vkolomeyko ah, good to see. I've added some further comments to that ticket.

**Pull request closed without merging**